src/modules/bulletproofs/bulletproofs_pp_norm_product_impl.h

```diff
@@ -0,0 +1,589 @@
+                secp256k1_gej_add_ge_var(&grj, &grj, &gs[G_GENS_LEN + i], NULL);
+                secp256k1_ge_set_gej_var(&gs[G_GENS_LEN + i/2], &grj);
+            }
+        }
+        g_len = g_len / 2;
+        h_len = h_len / 2;
+        r = q;
+        q = q_sq;
+        start_idx += 65;
+    }
+
+    secp256k1_scalar_get_b32(&proof[start_idx], &ns[0]);
+    secp256k1_scalar_get_b32(&proof[start_idx + 32], &ls[0]);
+    *proof_len = start_idx + 64;
+    secp256k1_scratch_apply_checkpoint(&ctx->error_callback, scratch, scratch_checkpoint);
+    return 1;
+}
+
+typedef struct ec_mult_verify_cb_data1 {
+    const unsigned char *proof;
+    const secp256k1_ge *commit;
+    const secp256k1_scalar *challenges;
+} ec_mult_verify_cb_data1;
+
+static int ec_mult_verify_cb1(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *cbdata) {
+    ec_mult_verify_cb_data1 *data = (ec_mult_verify_cb_data1*) cbdata;
+    if (idx == 0) {
+        *pt = *data->commit;
+        secp256k1_scalar_set_int(sc, 1);
+        return 1;
+    }
+    idx -= 1;
+    if (idx % 2 == 0) {
+        unsigned char pk_buf[33];
+        idx /= 2;
+        *sc = data->challenges[idx];
+        pk_buf[0] = 2 | (data->proof[65*idx] >> 1);
+        memcpy(&pk_buf[1], &data->proof[65*idx + 1], 32);
+        if (!secp256k1_eckey_pubkey_parse(pt, pk_buf, sizeof(pk_buf))) {
+            return 0;
+        }
+    } else {
+        unsigned char pk_buf[33];
+        secp256k1_scalar neg_one;
+        idx /= 2;
+        secp256k1_scalar_set_int(&neg_one, 1);
+        secp256k1_scalar_negate(&neg_one, &neg_one);
+        *sc = data->challenges[idx];
+        secp256k1_scalar_sqr(sc, sc);
+        secp256k1_scalar_add(sc, sc, &neg_one); /* Ignore overflow */
+        pk_buf[0] = 2 | data->proof[65*idx];
+        memcpy(&pk_buf[1], &data->proof[65*idx + 33], 32);
+        if (!secp256k1_eckey_pubkey_parse(pt, pk_buf, sizeof(pk_buf))) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+typedef struct ec_mult_verify_cb_data2 {
+    const secp256k1_scalar *s_g;
+    const secp256k1_scalar *s_h;
+    const secp256k1_ge *g_vec;
+    size_t g_vec_len;
+} ec_mult_verify_cb_data2;
+
+static int ec_mult_verify_cb2(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *cbdata) {
+    ec_mult_verify_cb_data2 *data = (ec_mult_verify_cb_data2*) cbdata;
+    if (idx < data->g_vec_len) {
+        *sc = data->s_g[idx];
+    } else {
+        *sc = data->s_h[idx - data->g_vec_len];
+    }
+    *pt = data->g_vec[idx];
+    return 1;
+}
+
+/* Verify the proof */
+int secp256k1_bulletproofs_pp_rangeproof_norm_product_verify(
+    const secp256k1_context* ctx,
+    secp256k1_scratch_space* scratch,
+    unsigned char* proof,
+    size_t proof_len,
+    unsigned char *transcript_hash32,
+    const secp256k1_scalar* r_ch,
+    secp256k1_bulletproofs_generators* g_vec,
+    size_t g_len,
+    secp256k1_scalar* c_vec,
+    size_t c_vec_len,
+    const secp256k1_ge* commit
```

Comment:
This argument is removed in the follow-up PR and it's unnecessary. Above `norm_produce_verify` assumes that `r` and `c_vec` (as @Liam-Eagen pointed out) are already included in `transcript_hash`, so it could also just assume that `commit` is in the transcript hash.

It'd be really great if we could clearly document the assumptions we put on the argument of the function with respect to the transcript because this is a common source of vulnerabilities. It will also make it easier to review the follow-up PR.

Reply:
> It'd be really great if we could clearly document the assumptions we put on the argument of the function with respect to the transcript because this is a common source of vulnerabilities. It will also make it easier to review the follow-up PR.

My initial goal was to create a stand-alone norm argument protocol and compose it with range proofs API. However, as I have slowly realized through this review, these are so easily composable for efficiency reasons. If we want to standalone norm argument, we should add the commitment `C` to transcript-hash because that is the initial input to the norm argument.

But if we use it in a parent protocol, the prover needs to invest additional resources to compute the final commitment only to commit it in the transcript. This is unnecessary as the final commit is a deterministic function of already committed points and is not shared in prover verifier interaction during the norm argument. 

To elaborate, there are three protocols:

1) `norm_arg_prove_internal` :  This does not commit to initial commitment points and assumes that the parent protocol has committed to it. We **want** `secp256k1_bulletproofs_pp_rangeproof_norm_product_prove` to do this.
2) `norm_arg_prove`: This should be a standalone function that we can use for norm argument. This would be the wrapper around `norm_arg_prove_internal` that adds commitment points, generators, etc to the transcript. The current version of `secp256k1_bulletproofs_pp_rangeproof_norm_product_prove` does this instead.
3) `rangeproof_prove`: This should use the (2) as a sub-protocol, not (1). The second PR changes this function to do (2) instead of (1). 

I think I will change the PR to 
- Change the secp256k1_bulletproofs_pp_rangeproof_norm_product_prove to follow (1)
- add the `norm_arg_prove`, `norm_arg_verify` and `norm_arg_commit` according to protocol (2) in tests as wrappers around (1)

Reply:
> Change the secp256k1_bulletproofs_pp_rangeproof_norm_product_prove to follow (1)

I agree.

> add the norm_arg_prove, norm_arg_verify and norm_arg_commit according to protocol (2) in tests as wrappers around (1)

I like this idea.